moveit_core/collision_detection/src/world.cpp

```diff
@@ -225,9 +225,14 @@ bool World::moveObject(const std::string& object_id, const Eigen::Isometry3d& tr
   if (transform.isApprox(Eigen::Isometry3d::Identity()))
     return true;  // object already at correct location
   ensureUnique(it->second);
-  for (size_t i = 0, n = it->second->shapes_.size(); i < n; ++i)
+  it->second->pose_ = transform * it->second->pose_;
+  for (Eigen::Isometry3d shape_pose : it->second->shape_poses_)
   {
-    it->second->shape_poses_[i] = transform * it->second->shape_poses_[i];
+    shape_pose = transform * shape_pose;
+  }
+  for (std::pair<std::string, Eigen::Isometry3d> pair : it->second->subframe_poses_)
```

Comment:
This should also be by reference.

Reply:
Real nice catch on both btw, thanks!

Reply:
Actually changing this one gives a compiler error:

```shell
/root/linked_ws_moveit/src/moveit/moveit_core/collision_detection/src/world.cpp: In member function ‘bool collision_detection::World::moveObject(const string&, const Isometry3d&)’:
/root/linked_ws_moveit/src/moveit/moveit_core/collision_detection/src/world.cpp:233:70: error: conversion from ‘std::pair<const std::__cxx11::basic_string<char>, Eigen::Transform<double, 3, 1> >’ to non-scalar type ‘std::pair<std::__cxx11::basic_string<char>, Eigen::Transform<double, 3, 1>&>’ requested
   for (std::pair<std::string, Eigen::Isometry3d&> pair : it->second->subframe_poses_)
                                                                      ^~~~~~~~~~~~~~~
```

Reply:
I think that's because you placed the reference next to the Isometry3d and not on the whole pair.  This is what you want I think:

```for (std::pair<std::string, Eigen::Isometry3d>& pair : it->second->subframe_poses_)```

Reply:
That gave me this error:
```shell
/root/linked_ws_moveit/src/moveit/moveit_core/collision_detection/src/world.cpp: In member function ‘bool collision_detection::World::moveObject(const string&, const Isometry3d&)’:
/root/linked_ws_moveit/src/moveit/moveit_core/collision_detection/src/world.cpp:233:70: error: invalid initialization of reference of type ‘std::pair<std::__cxx11::basic_string<char>, Eigen::Transform<double, 3, 1> >&’ from expression of type ‘std::pair<const std::__cxx11::basic_string<char>, Eigen::Transform<double, 3, 1> >’
   for (std::pair<std::string, Eigen::Isometry3d>& pair : it->second->subframe_poses_)
                                                                      ^~~~~~~~~~~~~~~
```
But this compiles:
`for (std::pair<const std::string, Eigen::Isometry3d>& pair : it->second->subframe_poses_)`

Reply:
Ah.  That makes sense that the string is const.  This is why I like using auto& in range based for loops.  It is often confusing to figure out there right type and the code doesn't really depend on the exact type, only that I know it is a pair of string to isometry.

Reply:
Yeah, it's comfortable when prototyping, but I actually found it difficult to debug later on when I didn't know what the types were. It's definitely more descriptive, so I'm in favor of the guideline.